psqlgraph/node.py

```diff
@@ -204,6 +215,58 @@ def edge_predicate(e):
                     queue.append((n, depth + 1))
                     marked.add(n.node_id)
 
+    def _dfs(self, edge_predicate=None, max_depth=None, edge_pointer="in"):
+        """
+        Perform a DFS, with `self` being the root node
+
+        To implement max depth, some node are visited more than once to update shortest
+        path. But those node should only be yield once.
+
+        :param edge_predicate: a predicate performed on an `edge` object in
+            order to decided whether to walk that edge or not
+        :type edge_predicate: func
+        :param max_depth: maximum distance to traverse
+        :type max_depth: int
+        :param edge_pointer: possible values `in`, `out`
+                            `in`: use node.edges_in, default behavior
+                            `out`: use edges_out
+        :type edge_pointer: str
+
+        :return: generator
+        """
+        StackItem = namedtuple('StackItem', ['node', 'next_child', 'level'])
+        edge_predicate = edge_predicate if callable(edge_predicate) else lambda _: True
+        max_depth = float('inf') if max_depth is None else max_depth
+
+        visited = {self.node_id: 0}
+        yield self
+        stack = [StackItem(self, 0, 0)]
+
+        while stack:
+            node, next_child, level = stack.pop()
+            if level >= max_depth:
+                continue
+            edges = node.edges_out if edge_pointer == "out" else node.edges_in
+
+            for j in range(next_child, len(edges)):
+                edge = edges[j]
+                n = edge.dst if edge_pointer == "out" else edge.src
+                if not edge_predicate(edge):
+                    continue
```

Comment:
Gets rid of some of the nesting if you use list comprehension 
```suggestion
            edges = node.edges_out if edge_pointer == "out" else node.edges_in
            edges = (edges[i] for i in range(next_child, len(edges))
            edges = (edge for edge in edges if not edge_predicate(edge))
            
            for edge in edges:
                n = edge.dst if edge_pointer == "out" else edge.src
```

Reply:
Need to record the index number j later.

Reply:
True. Updated to reflect. But again just a suggestion as below.

Reply:
@qiaouchicago Indeed! Using a generator or even chaining generators to filter/map and other operations before actually entering the for loop is a really cheap way to clean up code because its not evaluated until its actually iterated. Just remember to only force evaluation at the last possible moment cuz that is what actually gets expensive.